materialize-databricks/staged_file.go

```diff
@@ -0,0 +1,276 @@
+// to complete before returning.
+type stagedFile struct {
+	cols     []*sql.Column
+
+	// Random string that will serve as the directory for local files of this binding.
+	uuid string
+
+	// The full directory path of local files for this binding formed by joining tempdir and uuid.
+	dir string
+
+	// The remote root directory for uploading files
+	root string
+
+	filesAPI *files.FilesAPI
+
+	// Indicates if the stagedFile has been initialized for this transaction yet. Set `true` by
+	// start() and `false` by flush().
+	started bool
+
+	// Index of the current file for this transaction. Starts at 0 and is incremented by 1 for each
+	// new file that is created.
+	fileIdx int
+
+	// References to the current file being written.
+	buf     *fileBuffer
+	encoder *sql.CountingEncoder
+
+	// List of file names uploaded during the current transaction for transaction data, not
+	// including the manifest file name itself. These data file names randomly generated UUIDs.
+	uploaded []string
+
+	// Per-transaction coordination.
+	putFiles chan string
+	group    *errgroup.Group
+	groupCtx context.Context // Used to check for group cancellation upon the worker returning an error.
+}
+
+func newStagedFile(filesAPI *files.FilesAPI, root string, cols []*sql.Column) *stagedFile {
+	uuid := uuid.NewString()
+	var tempdir = os.TempDir()
+
+	return &stagedFile{
+		cols: cols,
+		uuid: uuid,
+		dir:  filepath.Join(tempdir, uuid),
+		root: root,
+		filesAPI: filesAPI,
+	}
+}
+
+func (f *stagedFile) start(ctx context.Context) error {
+	if f.started {
+		return nil
+	}
+	f.started = true
+
+	// Create the local working directory for this binding. As a simplification we will always
+	// remove and re-create the directory since it will already exist for transactions beyond the
+	// first one.
+	if err := os.RemoveAll(f.dir); err != nil {
+		return fmt.Errorf("clearing temp dir: %w", err)
+	} else if err := os.Mkdir(f.dir, 0700); err != nil {
+		return fmt.Errorf("creating temp dir: %w", err)
+	}
+
+	// Reset values used per-transaction.
+	f.uploaded = []string{}
+	f.group, f.groupCtx = errgroup.WithContext(ctx)
+	f.putFiles = make(chan string)
+
+	for i := 0; i < uploadConcurrency; i++ {
+		// Start the putWorker for this transaction.
+		f.group.Go(func() error {
+			return f.putWorker(f.groupCtx, f.putFiles)
+		})
+	}
+
+	return nil
+}
+
+func (f *stagedFile) encodeRow(row []interface{}) error {
+	// May not have an encoder set yet if the previous encodeRow() resulted in flushing the current
+	// file, or for the very first call to encodeRow().
+	if f.encoder == nil {
+		if err := f.newFile(); err != nil {
+			return err
+		}
+	}
+
+	if len(row) != len(f.cols) { // Sanity check
+		return fmt.Errorf("number of headers in row to encode (%d) differs from number of configured headers (%d)", len(row), len(f.cols))
+	}
+
+	d := make(map[string]interface{})
+	for idx := range row {
+		if v, ok := row[idx].(json.RawMessage); ok {
+			d[f.cols[idx].Field] = string(v)
```

Comment:
@williamhbaker this is already part of the dialect's element converter, and #1066 is merged and I have rebased, but I believe we still do not apply the dialect's converter when encoding staged files.

Reply:
Where we do `encodeRow`, we always (I think) do a `ConvertKey` or `ConvertAll`, which uses the dialect's converters, ex [here](https://github.com/estuary/connectors/blob/4e14682e4e1958deeb8e7fd1207350b8d7235e67/materialize-databricks/driver.go#L440-L444) and [here](https://github.com/estuary/connectors/blob/4e14682e4e1958deeb8e7fd1207350b8d7235e67/materialize-databricks/driver.go#L545-L552).

As it is now, the dialect is converting everything with a `FlatType` of `ARRAY`, `OBJECT`, or `MULTIPLE` to a `json.RawMessage` via [this](https://github.com/estuary/connectors/blob/4e14682e4e1958deeb8e7fd1207350b8d7235e67/materialize-databricks/sqlgen.go#L19C74-L19C96), and then here it is converting that to a string.

Anyway, what is here does work since it is an array of values (now? I can't remember if it was an object before and that is where my original comment about #1066 came from) - I would still recommend doing the conversion all the way to a `string` in the dialect's element converter; BigQuery does this for example for arrays and objects, unless there's something else I'm missing.